File: standalone_demo.py
#!/usr/bin/env python3
"""
Standalone Autodesk MCP Server Demo
Run this to see the MCP tools in action without needing Claude Desktop
"""

import json
import asyncio
from autodesk_mcp_server import (
    list_tools,
    call_tool,
    SAMPLE_MODELS
)

def print_header(text):
    """Print a nice header"""
    print("\n" + "=" * 70)
    print(f"  {text}")
    print("=" * 70 + "\n")

def print_tool(tool):
    """Print tool information"""
    print(f"📋 {tool.name}")
    print(f"   {tool.description}")
    print()

async def demo_list_tools():
    """Demo: Show all available MCP tools"""
    print_header("DEMO 1: Available MCP Tools")
    
    tools = await list_tools()
    print(f"Found {len(tools)} MCP tools:\n")
    
    for i, tool in enumerate(tools, 1):
        print(f"{i}. {tool.name}")
        print(f"   → {tool.description[:80]}...")
    
    print(f"\n✅ All {len(tools)} tools are registered and ready!")

async def demo_list_models():
    """Demo: List available Autodesk models"""
    print_header("DEMO 2: List Available Models")
    
    result = await call_tool("list_available_models", {
        "file_type_filter": "all"
    })
    
    data = json.loads(result[0].text)
    print(f"📁 Total models: {data['total_models']}\n")
    
    for model in data['models']:
        print(f"  • {model['filename']}")
        print(f"    Type: {model['type']} | Project: {model['project']}")
    
    print(f"\n✅ Successfully listed all models!")

async def demo_explore_model():
    """Demo: Explore model data in detail"""
    print_header("DEMO 3: Explore Office Building Model")
    
    result = await call_tool("explore_model_data", {
        "filename": "office-building-01.rvt",
        "query_type": "summary"
    })
    
    data = json.loads(result[0].text)
    print(f"🏢 {data['filename']}")
    print(f"\n{data['summary']}\n")
    print("Key Metrics:")
    
    metrics = data['key_metrics']
    print(f"  • Total Area: {metrics.get('total_area_sqft', 0):,} sq ft")
    print(f"  • Floors: {metrics.get('floor_count', 0)}")
    print(f"  • Rooms: {metrics.get('room_count', 0)}")
    print(f"  • Doors: {metrics.get('door_count', 0)}")
    print(f"  • Windows: {metrics.get('window_count', 0)}")
    print(f"  • Total Components: {metrics.get('component_count', 0):,}")
    
    print(f"\n✅ Successfully explored model data!")

async def demo_calculate_materials():
    """Demo: Calculate material quantities"""
    print_header("DEMO 4: Calculate Materials for Warehouse")
    
    result = await call_tool("calculate_material_takeoff", {
        "filename": "warehouse-facility-03.rvt",
        "material_type": "all"
    })
    
    data = json.loads(result[0].text)
    print(f"🏗️  {data['filename']}\n")
    print("Material Takeoff:")
    
    for material, details in data['material_takeoff'].items():
        print(f"\n  {material.upper()}:")
        for key, value in details.items():
            print(f"    • {key}: {value:,}")
    
    print(f"\n✅ Successfully calculated material quantities!")

async def demo_compare_models():
    """Demo: Compare two models"""
    print_header("DEMO 5: Compare Office Building vs Residential Home")
    
    result = await call_tool("compare_model_versions", {
        "model_a": "office-building-01.rvt",
        "model_b": "residential-home-02.rvt",
        "comparison_type": "quantities"
    })
    
    data = json.loads(result[0].text)
    print(f"📊 {data['comparison']}\n")
    print("Key Differences:")
    
    for key, diff in data['differences'].items():
        print(f"\n  {key}:")
        print(f"    Office Building: {diff['model_a']:,}")
        print(f"    Residential Home: {diff['model_b']:,}")
        print(f"    Difference: {diff['difference']:,} ({diff['percent_change']:+.1f}%)")
    
    print(f"\n✅ Successfully compared models!")

async def demo_generate_report():
    """Demo: Generate a comprehensive report"""
    print_header("DEMO 6: Generate Cost Estimate Report")
    
    result = await call_tool("generate_model_report", {
        "filename": "office-building-01.rvt",
        "report_type": "cost_estimate",
        "format": "json"
    })
    
    data = json.loads(result[0].text)
    print(f"💰 {data['report_title']}\n")
    print(f"Estimated Total Cost: ${data['estimated_total_cost_usd']:,.2f}")
    print(f"Cost per Sq Ft: ${data['cost_per_sqft']:,.2f}")
    print(f"\nAssumptions: {data['assumptions']}")
    
    print(f"\n✅ Successfully generated cost estimate!")

async def demo_search_components():
    """Demo: Search for components"""
    print_header("DEMO 7: Search for Windows in Office Building")
    
    result = await call_tool("search_model_components", {
        "filename": "office-building-01.rvt",
        "search_term": "window",
        "search_category": "all"
    })
    
    data = json.loads(result[0].text)
    print(f"🔍 Search: '{data['search_term']}' in {data['filename']}\n")
    print(f"Results Found: {data['results_found']}")
    
    for component in data['components']:
        print(f"\n  • {component['category']}")
        print(f"    Count: {component['count']}")
    
    print(f"\n✅ Successfully searched components!")

async def demo_get_help():
    """Demo: Get Autodesk help content"""
    print_header("DEMO 8: Get Help for Revit Families")
    
    result = await call_tool("get_autodesk_help", {
        "product": "revit",
        "topic": "families"
    })
    
    print(result[0].text)
    
    print(f"\n✅ Successfully retrieved help content!")

async def demo_complex_workflow():
    """Demo: Complex multi-step workflow"""
    print_header("DEMO 9: Complex Workflow - Analyze All Buildings")
    
    print("Step 1: List all models...")
    list_result = await call_tool("list_available_models", {"file_type_filter": "Revit"})
    models_data = json.loads(list_result[0].text)
    revit_models = [m['filename'] for m in models_data['models']]
    print(f"   Found {len(revit_models)} Revit models\n")
    
    print("Step 2: Analyze each model...")
    analyses = []
    for filename in revit_models:
        result = await call_tool("explore_model_data", {
            "filename": filename,
            "query_type": "summary"
        })
        data = json.loads(result[0].text)
        area = data['key_metrics'].get('total_area_sqft', 0)
        components = data['key_metrics'].get('component_count', 0)
        analyses.append({
            'filename': filename,
            'area': area,
            'components': components
        })
        print(f"   ✓ {filename}: {area:,} sq ft")
    
    print("\nStep 3: Calculate cost-effectiveness...")
    cost_results = []
    for analysis in analyses:
        result = await call_tool("generate_model_report", {
            "filename": analysis['filename'],
            "report_type": "cost_estimate",
            "format": "json"
        })
        data = json.loads(result[0].text)
        cost_per_sqft = data['estimated_total_cost_usd'] / analysis['area'] if analysis['area'] > 0 else 0
        cost_results.append({
            'filename': analysis['filename'],
            'cost_per_sqft': cost_per_sqft,
            'total_cost': data['estimated_total_cost_usd']
        })
    
    print("\nStep 4: Compare and recommend...")
    best = min(cost_results, key=lambda x: x['cost_per_sqft'])
    
    print(f"\n🏆 RECOMMENDATION:")
    print(f"   Most cost-effective: {best['filename']}")
    print(f"   Cost per sq ft: ${best['cost_per_sqft']:.2f}")
    print(f"   Total estimated cost: ${best['total_cost']:,.2f}")
    
    print(f"\n✅ Complex workflow completed successfully!")
    print("   This demonstrates AI orchestrating multiple MCP tools!")

async def run_all_demos():
    """Run all demos in sequence"""
    print("\n")
    print("╔════════════════════════════════════════════════════════════════════╗")
    print("║                                                                    ║")
    print("║          🏗️  AUTODESK MCP SERVER - STANDALONE DEMO               ║")
    print("║                                                                    ║")
    print("║     Demonstrating Model Context Protocol Tools in Action          ║")
    print("║              (No Claude Desktop Required!)                        ║")
    print("║                                                                    ║")
    print("╚════════════════════════════════════════════════════════════════════╝")
    
    demos = [
        demo_list_tools,
        demo_list_models,
        demo_explore_model,
        demo_calculate_materials,
        demo_compare_models,
        demo_generate_report,
        demo_search_components,
        demo_get_help,
        demo_complex_workflow,
    ]
    
    for i, demo in enumerate(demos, 1):
        try:
            await demo()
            if i < len(demos):
                input("\n👉 Press ENTER for next demo...")
        except Exception as e:
            print(f"\n❌ Error in demo: {e}")
            continue
    
    print_header("🎉 ALL DEMOS COMPLETED!")
    print("This demonstrates:")
    print("  ✅ 7 MCP tools working correctly")
    print("  ✅ Context-aware Autodesk workflows")
    print("  ✅ Multi-step AI orchestration")
    print("  ✅ Production-ready implementation")
    print("\n🚀 Ready to present to your audience!")
    print("\n")

def main():
    """Main entry point"""
    try:
        asyncio.run(run_all_demos())
    except KeyboardInterrupt:
        print("\n\n⚠️  Demo interrupted by user")
    except Exception as e:
        print(f"\n\n❌ Error: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()

File: standalone-demo.js
#!/usr/bin/env node

/**
 * Standalone Autodesk MCP Server Demo
 * Run this to see the MCP tools in action without needing Claude Desktop
 */

import { SAMPLE_MODELS } from "./sample-data.js";
import {
  listTools,
  callTool
} from "./tool-handlers.js";
import readline from 'readline';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function prompt(question) {
  return new Promise((resolve) => {
    rl.question(question, resolve);
  });
}

function printHeader(text) {
  console.log("\n" + "=".repeat(70));
  console.log(`  ${text}`);
  console.log("=".repeat(70) + "\n");
}

async function demo1_ListTools() {
  printHeader("DEMO 1: Available MCP Tools");
  
  const tools = listTools();
  console.log(`Found ${tools.length} MCP tools:\n`);
  
  tools.forEach((tool, i) => {
    console.log(`${i + 1}. ${tool.name}`);
    console.log(`   → ${tool.description.substring(0, 80)}...`);
  });
  
  console.log(`\n✅ All ${tools.length} tools are registered and ready!`);
}

async function demo2_ListModels() {
  printHeader("DEMO 2: List Available Models");
  
  const result = await callTool("list_available_models", {
    file_type_filter: "all"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`📁 Total models: ${data.total_models}\n`);
  
  data.models.forEach(model => {
    console.log(`  • ${model.filename}`);
    console.log(`    Type: ${model.type} | Project: ${model.project}`);
  });
  
  console.log(`\n✅ Successfully listed all models!`);
}

async function demo3_ExploreModel() {
  printHeader("DEMO 3: Explore Office Building Model");
  
  const result = await callTool("explore_model_data", {
    filename: "office-building-01.rvt",
    query_type: "summary"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`🏢 ${data.filename}`);
  console.log(`\n${data.summary}\n`);
  console.log("Key Metrics:");
  
  const metrics = data.key_metrics;
  console.log(`  • Total Area: ${(metrics.total_area_sqft || 0).toLocaleString()} sq ft`);
  console.log(`  • Floors: ${metrics.floor_count || 0}`);
  console.log(`  • Rooms: ${metrics.room_count || 0}`);
  console.log(`  • Doors: ${metrics.door_count || 0}`);
  console.log(`  • Windows: ${metrics.window_count || 0}`);
  console.log(`  • Total Components: ${(metrics.component_count || 0).toLocaleString()}`);
  
  console.log(`\n✅ Successfully explored model data!`);
}

async function demo4_CalculateMaterials() {
  printHeader("DEMO 4: Calculate Materials for Warehouse");
  
  const result = await callTool("calculate_material_takeoff", {
    filename: "warehouse-facility-03.rvt",
    material_type: "all"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`🏗️  ${data.filename}\n`);
  console.log("Material Takeoff:");
  
  for (const [material, details] of Object.entries(data.material_takeoff)) {
    console.log(`\n  ${material.toUpperCase()}:`);
    for (const [key, value] of Object.entries(details)) {
      console.log(`    • ${key}: ${value.toLocaleString()}`);
    }
  }
  
  console.log(`\n✅ Successfully calculated material quantities!`);
}

async function demo5_CompareModels() {
  printHeader("DEMO 5: Compare Office Building vs Residential Home");
  
  const result = await callTool("compare_model_versions", {
    model_a: "office-building-01.rvt",
    model_b: "residential-home-02.rvt",
    comparison_type: "quantities"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`📊 ${data.comparison}\n`);
  console.log("Key Differences:");
  
  for (const [key, diff] of Object.entries(data.differences)) {
    console.log(`\n  ${key}:`);
    console.log(`    Office Building: ${diff.model_a.toLocaleString()}`);
    console.log(`    Residential Home: ${diff.model_b.toLocaleString()}`);
    console.log(`    Difference: ${diff.difference.toLocaleString()} (${diff.percent_change >= 0 ? '+' : ''}${diff.percent_change}%)`);
  }
  
  console.log(`\n✅ Successfully compared models!`);
}

async function demo6_GenerateReport() {
  printHeader("DEMO 6: Generate Cost Estimate Report");
  
  const result = await callTool("generate_model_report", {
    filename: "office-building-01.rvt",
    report_type: "cost_estimate",
    format: "json"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`💰 ${data.report_title}\n`);
  console.log(`Estimated Total Cost: $${data.estimated_total_cost_usd.toLocaleString()}`);
  console.log(`Cost per Sq Ft: $${data.cost_per_sqft.toLocaleString()}`);
  console.log(`\nAssumptions: ${data.assumptions}`);
  
  console.log(`\n✅ Successfully generated cost estimate!`);
}

async function demo7_SearchComponents() {
  printHeader("DEMO 7: Search for Windows in Office Building");
  
  const result = await callTool("search_model_components", {
    filename: "office-building-01.rvt",
    search_term: "window",
    search_category: "all"
  });
  
  const data = JSON.parse(result.content[0].text);
  console.log(`🔍 Search: '${data.search_term}' in ${data.filename}\n`);
  console.log(`Results Found: ${data.results_found}`);
  
  data.components.forEach(component => {
    console.log(`\n  • ${component.category}`);
    console.log(`    Count: ${component.count}`);
  });
  
  console.log(`\n✅ Successfully searched components!`);
}

async function demo8_GetHelp() {
  printHeader("DEMO 8: Get Help for Revit Families");
  
  const result = await callTool("get_autodesk_help", {
    product: "revit",
    topic: "families"
  });
  
  console.log(result.content[0].text);
  
  console.log(`\n✅ Successfully retrieved help content!`);
}

async function demo9_ComplexWorkflow() {
  printHeader("DEMO 9: Complex Workflow - Analyze All Buildings");
  
  console.log("Step 1: List all models...");
  const listResult = await callTool("list_available_models", { 
    file_type_filter: "Revit" 
  });
  const modelsData = JSON.parse(listResult.content[0].text);
  const revitModels = modelsData.models.map(m => m.filename);
  console.log(`   Found ${revitModels.length} Revit models\n`);
  
  console.log("Step 2: Analyze each model...");
  const analyses = [];
  for (const filename of revitModels) {
    const result = await callTool("explore_model_data", {
      filename,
      query_type: "summary"
    });
    const data = JSON.parse(result.content[0].text);
    const area = data.key_metrics.total_area_sqft || 0;
    const components = data.key_metrics.component_count || 0;
    analyses.push({ filename, area, components });
    console.log(`   ✓ ${filename}: ${area.toLocaleString()} sq ft`);
  }
  
  console.log("\nStep 3: Calculate cost-effectiveness...");
  const costResults = [];
  for (const analysis of analyses) {
    const result = await callTool("generate_model_report", {
      filename: analysis.filename,
      report_type: "cost_estimate",
      format: "json"
    });
    const data = JSON.parse(result.content[0].text);
    const costPerSqft = analysis.area > 0 ? data.estimated_total_cost_usd / analysis.area : 0;
    costResults.push({
      filename: analysis.filename,
      cost_per_sqft: costPerSqft,
      total_cost: data.estimated_total_cost_usd
    });
  }
  
  console.log("\nStep 4: Compare and recommend...");
  const best = costResults.reduce((min, curr) => 
    curr.cost_per_sqft < min.cost_per_sqft ? curr : min
  );
  
  console.log(`\n🏆 RECOMMENDATION:`);
  console.log(`   Most cost-effective: ${best.filename}`);
  console.log(`   Cost per sq ft: $${best.cost_per_sqft.toFixed(2)}`);
  console.log(`   Total estimated cost: $${best.total_cost.toLocaleString()}`);
  
  console.log(`\n✅ Complex workflow completed successfully!`);
  console.log("   This demonstrates AI orchestrating multiple MCP tools!");
}

async function runAllDemos() {
  console.log("\n");
  console.log("╔════════════════════════════════════════════════════════════════════╗");
  console.log("║                                                                    ║");
  console.log("║          🏗️  AUTODESK MCP SERVER - STANDALONE DEMO               ║");
  console.log("║                                                                    ║");
  console.log("║     Demonstrating Model Context Protocol Tools in Action          ║");
  console.log("║              (No Claude Desktop Required!)                        ║");
  console.log("║                                                                    ║");
  console.log("╚════════════════════════════════════════════════════════════════════╝");
  
  const demos = [
    demo1_ListTools,
    demo2_ListModels,
    demo3_ExploreModel,
    demo4_CalculateMaterials,
    demo5_CompareModels,
    demo6_GenerateReport,
    demo7_SearchComponents,
    demo8_GetHelp,
    demo9_ComplexWorkflow,
  ];
  
  for (let i = 0; i < demos.length; i++) {
    try {
      await demos[i]();
      if (i < demos.length - 1) {
        await prompt("\n👉 Press ENTER for next demo...");
      }
    } catch (error) {
      console.log(`\n❌ Error in demo: ${error.message}`);
    }
  }
  
  printHeader("🎉 ALL DEMOS COMPLETED!");
  console.log("This demonstrates:");
  console.log("  ✅ 7 MCP tools working correctly");
  console.log("  ✅ Context-aware Autodesk workflows");
  console.log("  ✅ Multi-step AI orchestration");
  console.log("  ✅ Production-ready implementation");
  console.log("\n🚀 Ready to present to your audience!");
  console.log("\n");
  
  rl.close();
}

// Run all demos
runAllDemos().catch((error) => {
  console.error("\n\n❌ Error:", error.message);
  rl.close();
  process.exit(1);
});
